### ISSUE TITLE
No autocompletion for `inherent_associated_types` nightly feature

### ISSUE THREAD

**Issue by Ziothh** [rust-lang/rust-analyzer#16240]
As stated above: my LSP doens't give auto completion for a constant that's typed by the inherent_associated_types feature.
Rustc does compile the program when using the feature.

UPDATE: After some testing I found that this only happens for `const` values and not variables defined with `let` (regardless of scope).

# Code example
[My repo](https://github.com/Ziothh/gb-emu) is public so if my code examples aren't clear enough you can always take a look:
[Current commit when writing this issue](https://github.com/Ziothh/gb-emu/commit/61d2a1684228cfd9a7596206470f2f6d9e33a817). (Bumping the nightly version from 2023-11-19 to 2023-12-29 has nothing to do with this issue).

```rust
// /core/src/hardware/bus.rs
use std::ops;

pub struct MemoryBus(/* ... */);
impl MemoryBus {
    pub type Addr = u16;
    // Also tried replacing MemoryBus::Addr with just plain u16 but didn't work
    // Moving it outside impl and just doing a plain old type does work
    pub type Region = ops::RangeInclusive<MemoryBus::Addr>; 
}

pub mod regions {
    pub const OAM: MemoryBus::Region = 0xFE00..=0xFE9F
}
```
```rust
// /core/src/io/graphics/objects.rs
use crate::hardware::bus;

mod oam {
    struct ObjectAttributes<'a>(bytes: &'a [u8; 4]);

    impl<'a> ObjectAttributes<'a> {
        pub fn from_bus(&memory_bus:  &bus::MemoryBus) -> [ObjectAttributes; 40] {
             // Function I was working on while encountering this issue.
             bus::regions::OAM./* I'd want my LSP to provide me with auto completion here */
        }
    }
}
```

# Toolchain versions
## `rust-analyzer`
`$ rust-analyzer --version --verbose`
rust-analyzer 1.77.0-nightly (fb5ed72 2023-12-28)
## `rustc`
`$ rustc --version --verbose`
rustc 1.77.0-nightly (fb5ed726f 2023-12-28)
binary: rustc
commit-hash: fb5ed726f72c6d16c788517c60ec00d4564b9348
commit-date: 2023-12-28
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6

# Rust feature tracking issue
https://github.com/rust-lang/rust/issues/8995

**Comment by Veykril** [comment]
The feature is not supported at all in rust-analyzer, and iirc to implement it we need the new trait solver (or chalk support for it)

**Comment by Ziothh** [comment]
Hey @Veykril, thanks for the quick response.

Are there any relevant issues you could forward me too so I can keep up to date with progress being made on that trait solver you mentioned and/or chalk? 

Should I go ahead and close this issue or do you want to keep it open until the feature is supported?

**Comment by Veykril** [comment]
We can keep this open I don't think we have an issue tracking this feature here yet.

**Comment by flodiebold** [comment]
> actually that was assoc type default, unsure about this feature

Yeah, I feel like this shouldn't be particularly hard to support -- there's really not much difference between
```rust
impl Foo {
    type Bar = u32;
}
```
and
```rust
type FooBar = u32;
````
, is there?

**Comment by Ziothh** [comment]
Just to be clear: Foo::Bar seems to be supported by rust-analyzer. 

It just doesn't resolve (for auto completion & type inference) when a const that is typed with it and is used somewhere.
